core/ibft.go

```diff
@@ -666,12 +667,31 @@ func (i *IBFT) validateProposal(msg *proto.Message, view *proto.View) bool {
 		return false
 	}
 
+	if !messages.HasUniqueSenders(certificate.RoundChangeMessages) {
```

Comment:
Why we are return with `return false` if one dishonest node sends same message twice in a round? Shouldn't we just  remove his messages from slice and than continue with other checks? In that case, `HasQuorum` should be called after removing messages from duplicate senders.

Also, because round check is bellow, is it possible to have two messages by the same sender but for different rounds (maybe latency issue in network)? In that case, maybe we should not consider this as non valid condition.

Reply:
All valid points as far as I am concerned. Im inclined to to your opinion actually, but as far as I am concerned maybe I am grossly overlooking something within the paper to support your opinion.

We are looking at Alg2 : 71 of the paper. It is not doing any filtering or sorting. Last check clearly states _**any**_ message in RCC is signed by a different validator.

We are looking at a specific proposal here for a specific height and round > 0. We should not be seeing messages with different rounds here. If you want to round change, you are inherently sending r + 1 in RCC message. So to answer you question, i think they are not either/or, they all must hold true.

The reason why I might suggest we go this route instead of what you suggested (which I agree with) is because I don't know what I am missing and why is this not implemented in the paper like you said.